packages/react-router/docs/guides/testing.md

```diff
@@ -158,43 +99,51 @@ const App = () => (
     />
   </div>
 );
+```
 
-// the actual test!
-it("navigates around", done => {
-  renderTestSequence({
-    // tell it the subject you're testing
-    subject: App,
-
-    // and the steps to execute each time the location changes
-    steps: [
-      // initial render
-      ({ history, div }) => {
-        // assert the screen says what we think it should
-        console.assert(div.innerHTML.match(/Welcome/));
-
-        // now we can imperatively navigate as the test
-        history.push("/dashboard");
-      },
-
-      // second render from new location
-      ({ div }) => {
-        console.assert(div.innerHTML.match(/Dashboard/));
-
-        // or we can simulate clicks on Links instead of
-        // using history.push
-        Simulate.click(div.querySelector("#click-me"), {
-          button: 0
-        });
-      },
-
-      // final render
-      ({ location }) => {
-        console.assert(location.pathname === "/");
-        // you'll want something like `done()` so your test
-        // fails if you never make it here.
-        done();
-      }
-    ]
+```jsx
+// you can also use a renderer like "@testing-library/react" or "enzyme/mount" here
+import { render, unmountComponentAtNode } from "react-dom";
+import { act } from 'react-dom/test-utils';
+import { Router } from "react-router";
+import { createMemoryHistory } from "history";
+
+// app.test.js
+it("navigates home when you click the logo", async => {
+  // in a real test a renderer like "@testing-library/react"
+  // would take care of setting up the DOM elements
+  const root = document.createElement('div');
+  document.body.appendChild(root);
+  
+  const history = createMemoryHistory();
+  
+  // Set initial location
+  history.push('/my/initial/route');
```

Comment:
The history instance should be created with the initial location.
```js
// Create history with initial location
const history = createMemoryoHistory({
  initialEntries: ['/my/initial/route']
});
```

Reply:
Instead of using the `history` lib and `createMemoryHistory` I think I'd prefer the test to just use a `<MemoryRouter>`. That's what all our internal tests do.

Reply:
https://github.com/ReactTraining/react-router/pull/6903/files#r325358949

> That's what all our internal tests do.

I'd rather just show the thing that works at scale rather than the most convenient way for one-offs. In an app testing scenario you often have a shared wrapper around the render that sets up the critical context providers. This is not touched that often by app devs, so whatever is set up there needs to handle all use cases. If you can't access the location/history objects you create a need for workarounds.

Open to changing my mind if there is a clean way to handle use cases like checking redirects, query param updates, etc. with the MemoryRouter.

Reply:
Yep, you can do all that stuff with a memory router. It's just a different backend (an in-memory URL storage system instead of using the browser) but the `location` objects you get are the same for all routers.

So e.g. if you wanted to assert you got a specific query from a router you could do:

```js
let searchParams;

render(
  <MemoryRouter initialEntries={['/home']}>
    <Switch>
      <Route path="/home"/>
      <Route path="/products" render={({ location }) => {
        // Here you have access to the location, match (including URL path params), etc.
        searchParams = new URLSearchParams(location.search);
        return null;
      }}/>
  </MemoryRouter>
)

act(() => {
  // Click a <Link> to /products?id=1234
})

expect(searchParams.has('id')).toBe(true);
expect(searchParams.get('id')).toEqual('1234');
```

Reply:
Makes sense. I'm not sure I understand the aversion to using a regular BrowserRouter though 🤔

Reply:
Does MemoryRouter expose the location prop outside the React tree so you can reference it in assertions? That is the issue.

Reply:
I'm not opposed to using a `<BrowserRouter>`. I'm just trying to de-emphasize using the `history` API directly in tests. Instead, I'm suggesting sticking to the React API when making tests for React Router because it's only API that the router really gives you. The `history` library is just a dependency. Making assertions on it is basically making sure that a transitive dependency does its job, which I try to avoid when testing my app.

Reply:
I agree with you. Additionally, if we use the `BrowserRouter` then people should be able to assert on `window.location` if they feel like they need to, and they wouldn't need to interact with the history package.

Reply:
You can't clear the global history state between tests with BrowserRouter though, it only resets when the whole JSDOM env is set up (per file).

Reply:
Ah, that's an interesting point.

To sum up:

1. Sometimes people need to assert on the current URL because as Alex said in another comment "URLs are part of the API of the app"
2. To do this we'd need to either use BrowserRouter or create a custom history object (via `createMemoryHistory`)
3. If we use BrowserRouter, then the history state doesn't get reset between tests (and there is no way to reset history state manually I guess?)
4. If we use a custom history object, then we're exposing some implementation details of the router.

Please let me know if I'm missing something. The trade-offs feel equally trade-offy to me, so I'm not going to have a firm opinion either way on the example and I'll be happy with whatever the maintainers choose here.

Reply:
> You can't clear the global history state between tests with BrowserRouter though

This is typically why I use a `<MemoryHistory>`, so we can recreate the state of the history stack without too much work. It's a really nice abstraction that IMO is the easiest way to get started writing tests with the router.

Also, resetting the global state differs depending on the method that you're using for storing the URL in the browser. If you're using the HTML5 history API, you'll probably want to [`replaceState(null, null, '/')` before each test](https://github.com/ReactTraining/history/blob/3f69f9e07b0a739419704cffc3b3563133281548/modules/__tests__/BrowserHistory-test.js#L8-L13). But if you're using hash-based routing, you'll probably want to do `window.location.hash = ''` or `window.history.replaceState(null, null, '#/')` depending on which browsers you need to support. So there's some complexity here that I'd say is completely incidental and not really something you should care about testing. Just let the history lib handle it and read the `location` from whatever the router gives you.

In addition, I'm fairly certain there will be situations in the not-too-distant future where the router may give you a different `location` than what is actually in the address bar to accommodate routes that need to suspend while doing some data fetching.

So in general I'm trying to discourage people working directly with the `history` API and learning to work within the router API, including in their tests. We'll be shipping several hooks this week, including `useLocation()` which should make it easier to access and assert on the current `location` from anywhere you want.

Reply:
Note: replaceState doesn't clear the stack. You can still go back farther than the stack entry that got replaced.

Reply:
@alexkrolick Yes, I know :) But I never do that in tests because you don't know if e.g. `window.history.back()` will cause a full page reload or not.

Reply:
OK I changed the main example to MemoryRouter